lib/auth/streamingV4/trailingChecksumTransform.js

```diff
@@ -0,0 +1,92 @@
+const { Transform } = require('stream');
+const { errors } = require('arsenal');
+
+/**
+ * This class is designed to handle the chunks sent in a streaming
+ * v4 Auth request
+ */
+class TrailingChecksumTransform extends Transform {
+    /**
+     * @constructor
+     * @param {object} log - logger object
+     * @param {function} errCb - callback called if an error occurs
+     */
+    constructor(log, errCb) {
+        super({});
+        this.log = log;
+        this.errCb = errCb;
+        this.chunkSizeBuffer = Buffer.alloc(0);
+        this.outputBuffer = Buffer.alloc(0);
+        this.toFlush = Buffer.alloc(0);
+        this.bytesToRead = 0; // when a chunk is advertised, the size is put here and we forward all bytes
+        this.streamClosed = false;
+    }
+
+    /**
+     * This function will remove the trailing checksum from the stream
+     *
+     * @param {Buffer} chunkInput - chunk from request body
+     * @param {string} encoding - Data encoding
+     * @param {function} callback - Callback(err, justDataChunk, encoding)
+     * @return {function} executes callback with err if applicable
+     */
+    _transform(chunkInput, encoding, callback) {
+        let chunk = chunkInput;
+        for (;;) {
+            if (chunk.byteLength === 0) {
+                break;
+            }
+            if (this.streamClosed) {
+                return callback(null, '', encoding);
+            }
+            if (this.bytesToRead > 0) {
+                if (chunk.byteLength <= this.bytesToRead) {
+                    // chunk is smaller than the advertised size
+                    // forward the whole chunk
+                    this.bytesToRead -= chunk.byteLength;
+                    this.outputBuffer = Buffer.concat([this.outputBuffer, chunk]);
+                    break;
+                }
+                // chunk is bigger than the advertised size
+                this.outputBuffer = Buffer.concat([this.outputBuffer, chunk.subarray(0, this.bytesToRead)]);
+                chunk = chunk.subarray(this.bytesToRead);
+                this.bytesToRead = 0;
+                continue;
+            }
+            const lineBreakIndex = chunk.indexOf('\r\n');
```

Comment:
I think it needs to be:

```suggestion
const lineBreakIndex = this.chunkSizeBuffer.indexOf('\r\n');
```

Otherwise, it wouldn't be able to detect the newline with two chunks such as:

```
abcd\r  # chunk 1
\nefgh  # chunk 2
```

Then, you'll need to adapt the computed indices below with this change.